core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -227,7 +228,8 @@ public Balancer(Logger logger, RoutingAllocation allocation, WeightFunction weig
             this.routingNodes = allocation.routingNodes();
             this.metaData = allocation.metaData();
             avgShardsPerNode = ((float) metaData.getTotalNumberOfShards()) / routingNodes.size();
-            buildModelFromAssigned();
+            nodes = Collections.unmodifiableMap(buildModelFromAssigned());
+            sorter = newNodeSorter();
```

Comment:
sorter being a field now, it does not need to be pass around anymore, see e.g. `private String[] buildWeightOrderedIndices(NodeSorter sorter)`.